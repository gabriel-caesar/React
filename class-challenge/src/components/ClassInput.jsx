/* eslint-disable react/destructuring-assignment */
import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Count extends Component {
  render() {
    return <div>{this.props.count}</div>;
  }
}

class ClassInput extends Component {
  constructor(props) {
    super(props);

    this.state = {
      todos: ['Just some demo tasks', 'As an example'],
      inputVal: '',
      isEdit: false,
      editedVal: null,
      count: 2,
    };

    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.editInput = this.editInput.bind(this);
  }

  increment() {
    this.setState((prev) => ({ count: prev.count + 1 }));
  }

  decrement() {
    if (this.state.count <= 0) {
      return { count: 0 };
    } else {
      return this.setState((prev) => ({ count: prev.count - 1 }));
    }
  }

  handleInputChange(e) {
    this.setState((state) => ({
      ...state,
      inputVal: e.target.value,
    }));
  }

  handleSubmit(e) {
    e.preventDefault();
    this.increment();
    this.setState((state) => ({
      todos: state.todos.concat(state.inputVal),
      inputVal: '',
    }));
  }

  editInput(e) {
    return this.setState({
      editedVal: e.target.value,
    });
  }

  editButton(index) {
    // create a copy of the desired array to mutate
    const updatedTodos = this.state.todos;
    // modify the desired index element
    updatedTodos[index] = this.state.editedVal;
    // assign the real todos array to the updated copy of it
    this.setState({
      todos: updatedTodos,
      isEdit: null,
    });
  }

  
  render() {
    return (
      <section>
        {/* eslint-disable-next-line react/prop-types */}
        <h3>{this.props.name}</h3>
        {/* The input field to enter To-Do's */}
        <form onSubmit={this.handleSubmit}>
          {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
          <label htmlFor='task-entry'>Enter a task: </label>
          <input
            type='text'
            aria-label='task-entry'
            placeholder='enter your todo...'
            value={this.state.inputVal}
            onChange={this.handleInputChange}
          />
          <button type='submit'>Submit</button>
          <Count count={this.state.count} />
        </form>
        <h4>All the tasks!</h4>
        {/* The list of all the To-Do's, displayed */}
        <ul>
          {this.state.todos.map((todo, index) => (
            <div key={todo}>
              {this.state.isEdit !== index ? (
                <li>{todo}</li>
              ) : (
                <input
                  type='text'
                  data-testid={todo}        
                  // make sure you use the keyword 'this' for functions
                  onChange={e => this.editInput(e)}
                  // ensures input shows with the past value
                  value={
                    this.state.editedVal === null ? todo : this.state.editedVal
                  }
                />
              )}
              <button
                onClick={() => {
                  this.decrement();
                  this.setState(() => ({
                    todos: this.state.todos.filter((x) => x !== todo),
                  }));
                }}
              >
                Delete
              </button>

              {this.state.isEdit !== index ? (
                <button
                  data-testid={index}
                  onClick={() =>
                    this.setState({
                      isEdit: index,
                    })
                  }
                >
                  Edit
                </button>
              ) : (
                <button onClick={() => this.editButton(index)}>Resubmit</button>
              )}
            </div>
          ))}
        </ul>
      </section>
    );
  }
}

ClassInput.PropTypes = {
  name: PropTypes.string,
}

export default ClassInput;
