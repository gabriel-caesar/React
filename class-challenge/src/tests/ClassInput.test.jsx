import { beforeEach, describe, it } from 'vitest';
import { render, screen } from '@testing-library/react'; // (@) defines a scope of testing libraries which react is one of then
import { userEvent } from '@testing-library/user-event';
import ClassInput from '../components/ClassInput';

describe('User Interactions Test', () => {
  // setting up 'mock user' so that we can mock its interactions with the page
  const user = userEvent.setup();

  // it needs to be declared globally to be used down in the test functions
  let editButton;

  beforeEach(() => {
    render(<ClassInput />);
    editButton = screen.getByTestId('0');
  });

  it('Creates a snapshot for comparisons', () => {
    // snapshot running great, capturing the current HTML
    // output as a file to compare with future snapshots
    const { container } = render(<ClassInput />);
    expect(container).toMatchSnapshot();
  });

  it('Is the button text changing from (Submit) to (Resubmit)?', async () => {
    // awaiting user interaction (click the button)
    await user.click(editButton);

    // awaits the button 'Submit' to turn into 'Resubmit'
    const updatedButton = await screen.findByText(/resubmit/i);

    // expecting the button text to match Resubmit
    expect(editButton.textContent).toMatch(/resubmit/i);

    screen.debug(editButton);

    expect(updatedButton).toBeInTheDocument();
  });

  it('Is input showing the first task of the (todos) array?', async () => {
    // awaiting user interaction (click the button)
    await user.click(editButton);

    expect(
      screen.getByDisplayValue('Just some demo tasks')
    ).toBeInTheDocument();
  });

  it('Mocks a user creating a new task', async () => {
    await user.type(
      screen.getByPlaceholderText('enter your todo...'),
      'This is a test task'
    );

    await user.click(screen.getByRole('button', { name: 'Submit' }));

    expect(screen.getByText('This is a test task')).toBeInTheDocument();
    // showing debug log to reflect changes
    screen.debug();
  });
});
