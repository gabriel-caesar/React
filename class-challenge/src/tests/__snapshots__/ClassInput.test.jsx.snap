// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`User Interactions Test > Creates a snapshot for comparisons 1`] = `
<div>
  <section>
    <h3 />
    <form>
      <label
        for="task-entry"
      >
        Enter a task: 
      </label>
      <input
        aria-label="task-entry"
        placeholder="enter your todo..."
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Submit
      </button>
      <div>
        2
      </div>
    </form>
    <h4>
      All the tasks!
    </h4>
    <ul>
      <div>
        <li>
          Just some demo tasks
        </li>
        <button>
          Delete
        </button>
        <button
          data-testid="0"
        >
          Edit
        </button>
      </div>
      <div>
        <li>
          As an example
        </li>
        <button>
          Delete
        </button>
        <button
          data-testid="1"
        >
          Edit
        </button>
      </div>
    </ul>
  </section>
</div>
`;
